### ISSUE TITLE
ImagesAPIViewSet query params for rendition?

### ISSUE THREAD

**Issue by andreasnuesslein** [wagtail/wagtail#7973]
### Is your proposal related to a problem?

The problem is: it's not easy to seamlessly get renditions from the API without extra code within the `Page`s or `Block`s.

### Describe the solution you'd like

I'd like to just do something like this:
`"get" /api/v2/images/1/?renditions=max-1200x1200,height-200`

and the result would be something like this:
```
{
    "id": 1,
    "meta": {
        "type": "wagtailimages.Image",
        "detail_url": "http://localhost:9000/api/v2/images/1/",
        "tags": [],
        "download_url": "/media/original_images/processed-b7d09dd7-449a-4b96-9e1a-b3c471e4abad_m2p0kWLk_pcPaWKf.jpeg"
    },
    "renditions": {
        "max-1200x1200": "/media/....",
        "height-200": "/media/..."
    },
    "title": "processed-b7d09dd7-449a-4b96-9e1a-b3c471e4abad_m2p0kWLk",
    "width": 2522,
    "height": 3362
}
```

### Additional context

I would create a PR for this if there's interest :slightly_smiling_face:  :+1:

**Comment by gasman** [comment]
It sounds like what you're describing is essentially the same thing as the image serve view: https://docs.wagtail.org/en/stable/advanced_topics/images/image_serve_view.html

**Comment by andreasnuesslein** [comment]
I had a feeling you were gonna answer within 5 minutes :D

And I also figured something like this existed already. However it's not exactly what I'm looking for since either the user has to interact or I have to write extra code. 

When I create a `StructBlock` that has an `ImageChooserBlock`, I eventually get the ID of that image back via the API.
Now in my frontend, it would be nice to then just query for the rendition, which can be different depending on my frontend context. I.e. on PageA.svelte I want `fetch('/api/v2/images/1/?renditions=max-1200x1200')`, on PageB.svelte I want `fetch(/api/v2/images/1/?renditions=height-400)`.

Obviously if I only need a few different renditions I can just do that in my Block's `get_api_representation`, and the overhead is negligable.

I agree that my proposal has the flaw of any attacker being able to just flood the harddrive of any Wagtail-instance out there with renditions, however as an alternative proposal maybe we could just add a settings-variable that allows me to specify the renditions I might need?

But yeah: It's not a high-priority problem. Just an idea.

**Comment by kaedroho** [comment]
Thanks for the suggestion @andreasnuesslein 

The problem with implementing something like this in the Wagtail API is it would make it easy for someone to overload a site by sending it lots of requests for images at random sizes.

I think we would need to either limit the options for what rendition sizes can be requested on the server side, or require signed URLs on the client side.

Wagtail already has both options built in, but this could be documented better and there currently isn't a code snippet or library that you could use to generate the signed URLs for the image serve view.

**Comment by andreasnuesslein** [comment]
> I think we would need to either limit the options for what rendition sizes can be requested on the server side, or require signed URLs on the client side.

Yeah. If I understand correctly, signing would imply the client to have a SECRET, which is tricky to keep secret in case of a pure frontend package. So I was also thinking maybe just a settings-variable that could apply system-wide to any instance of `Image`s .

This way you'd just say "While building your frontend, you may specify a WAGTAIL_RENDITION_OPTIONS in settings so that any of these renditions might be requested at any time". In this case I'd even not do the whole GET-parameter stuff anymore but just deliver the rendition (options) within the response JSON directly

> ```
> {
>     "id": 1,
>     "meta": {
>         "type": "wagtailimages.Image",
>         "detail_url": "http://localhost:9000/api/v2/images/1/",
>         "tags": [],
>         "download_url": "/media/original_images/processed-b7d09dd7-449a-4b96-9e1a-b3c471e4abad_m2p0kWLk_pcPaWKf.jpeg"
>     },
>     "renditions": {
>         "max-1200x1200": "/media/....",
>         "height-200": "/media/..."
>     },
>     "title": "processed-b7d09dd7-449a-4b96-9e1a-b3c471e4abad_m2p0kWLk",
>     "width": 2522,
>     "height": 3362
> }
> ```

**Comment by andreasnuesslein** [comment]
And as an semi off-topic follow-up: For consistency reasons it would be helpful to have a simple Serializer/Endpoint that always delivers the same format for Images. Does something like that exist already and I haven't found it?

What I mean is: Right now I have a `CustomImage(AbstractImage)` which brings this method:
```python
    def get_api_representation(self, renditions=None):
        retval = {
            "id": self.id,
            "meta": {
                # "type": settings.WAGTAILIMAGES_IMAGE_MODEL,
                # "detail_url": ....
                "tags": self.tags.all(),
                "download_url": self.file.url,
            },
            "title": self.title,
            "width": self.width,
            "height": self.height,
            "caption": self.caption,
            "copyright": self.copyright,
        }
        if renditions:
            retval["renditions"] = {r: self.get_rendition(r).url for r in renditions}
        return retval
```

In my `StructBlock`s I then do
```
return { "picture": value.get("picture").get_api_representation(),...
  # or: .get_api_representation(["height-340"])
```

And for the Pages where I use my `CustomImage`, I have a custom serializer
```
class CustomImageSerializer(fields.Field):
    def to_internal_value(self, data):
        pass

    def to_representation(self, image):
        try:
            return image.get_api_representation()
        except SourceImageIOError:
            return {"error": "SourceImageIOError"}
```


Am I missing something that will simplify this? It feels quite boilerplate-y. I think that also inspired this ticket here.
